features/329-principal-apis/design.md

```diff
@@ -0,0 +1,100 @@
+# Background
+Auth gear supports various identity providers, such as password credentials and
+OAuth. Auth gear should support all identity providers equally well, and
+provide a consistent API for manipulating principals.
+
+
+# Use Cases
+- Get which method is used to log in the current session
+- List all methods to authenticated as a user
+- Add/remove user authentication method
+- Send welcome email to OAuth accounts
+- Populate user metadata according to external profile
+- Create password credentials for users signed up with OAuth
+
+
+# Glossary
+
+**Identity Provider**:
+A method to assert the identity of user. For example, providing a password or
+perform OAuth flow.
+
+**Principal**:
+An identity associated with user. A user can have multiple principals. Since
+the term 'Principal' is defined differently across many frameworks and
+standards, we will instead use a simpler term 'Identity' for public interfaces.
+
+
+# Proposed Design
```

Comment:
Can you add the function names related with each section/features below? Would make it easier to references with the api.md thanks!

Reply:
Further comments below, resolve this first.